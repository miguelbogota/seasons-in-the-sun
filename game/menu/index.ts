export { useHandleMenu } from './use-handle-menu.hook';
