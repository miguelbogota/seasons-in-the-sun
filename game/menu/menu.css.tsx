import { style, theme } from '@app/theme';

const menu = style({
  position: 'fixed',
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
  zIndex: 90,
  width: '100vw',
  height: '100vh',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  backgroundColor: theme.palette('background.sheet', { alpha: 0.5 }),
});

const button = style({
  padding: '2rem',
  backgroundColor: theme.palette('primary'),
  cursor: 'pointer',
  borderRadius: 5,
});

const loading = style({
  selectors: {
    [`${button}&`]: {
      backgroundColor: theme.palette('text.disabled'),
      cursor: 'default',
    },
  },
});

export const menuStyles = {
  menu,
  button,
  loading,
};
