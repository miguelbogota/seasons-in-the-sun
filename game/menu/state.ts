import { create as createStore } from 'zustand';

/**
 * Game menu state.
 */
export type GameMenuState = {
  /** Value to check if the menu is open or not (Initial values is `true`). */
  isOpen: boolean;
  /** Value to check if the menu is loading or not (Initial values is `false`). */
  isLoading: boolean;
  /** Function to change the loading state. */
  setLoading: (newState: boolean) => void;
  /** Function to open the menu. */
  open: () => void;
  /** Function to close the menu. */
  close: () => void;
  /** Function to toggle the menu. */
  toggle: () => void;
  /** Function to get the selector of the menu. */
  selector: (prefix?: boolean) => string;
};

/**
 * Game menu state using zustand.
 */
export const useGameMenuState = createStore<GameMenuState>((set) => ({
  isOpen: true,
  isLoading: false,
  setLoading: (newState) => set(() => ({ isLoading: newState })),
  open: () => set(() => ({ isOpen: true })),
  close: () => set(() => ({ isOpen: false })),
  toggle: () => set((state) => ({ isOpen: !state.isOpen })),
  selector: (prefix = false) => {
    const htmlSelector = 'play-button';
    return prefix ? `#${htmlSelector}` : htmlSelector;
  },
}));
