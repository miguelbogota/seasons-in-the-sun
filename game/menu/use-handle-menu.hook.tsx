import { Navigation } from '@app/website/navigation/navigation.component';
import { type FC, useEffect, useState } from 'react';

import { useGameMenu } from './store';

/** Hook to handle the initial setup for the menu. */
export function useHandleMenu() {
  const [loading, setLoading] = useState(true);
  const gameMenu = useGameMenu();

  useEffect(
    () => {
      if (gameMenu.isGameMenuOpen) {
        gameMenu.closeGameMenu();
      }

      const handleEscape = (event: KeyboardEvent) => {
        if (event.key === 'Escape') {
          gameMenu.toggleGameMenu();
        }
      };

      window.addEventListener('keydown', handleEscape);
      setLoading(false);

      return () => {
        gameMenu.openGameMenu();
        window.removeEventListener('keydown', handleEscape);
      };
    },
    // Disabled since it should only run once.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  const Menu = () => gameMenu.isGameMenuOpen && <Navigation />;

  return [loading, Menu] as [loading: boolean, Menu: FC];
}
