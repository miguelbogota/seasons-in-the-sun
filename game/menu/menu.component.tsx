import { Spinner } from '@app/lib/spinner';
import { Navigation } from '@app/website/navigation';
import clsx from 'clsx';

import { menuStyles } from './menu.css';
import { useGameMenuState } from './state';

/**
 * Game menu to show when the esc key is pressed (Or when the lock API from the PointerLockControls
 * gets triggered).
 */
export function GameMenu() {
  const isOpen = useGameMenuState((state) => state.isOpen);
  const isLoading = useGameMenuState((state) => state.isLoading);
  const selector = useGameMenuState((state) => state.selector);

  return (
    <div style={{ display: isOpen || isLoading ? 'block' : 'none' }}>
      <Navigation />
      <div className={menuStyles.menu}>
        <button
          id={selector()}
          className={clsx(menuStyles.button, { [menuStyles.loading]: isLoading })}
          disabled={isLoading}
        >
          {isLoading ? <Spinner /> : <h3>Click to play</h3>}
        </button>
      </div>
    </div>
  );
}
