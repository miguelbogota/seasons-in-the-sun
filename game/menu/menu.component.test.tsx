import { render } from '@testing-library/react';

import { GameMenu } from './menu.component';

describe('Game / Menu / Component', () => {
  it('should match snapshot', () => {
    const { container } = render(<GameMenu />);
    expect(container).toMatchSnapshot();
  });
});
