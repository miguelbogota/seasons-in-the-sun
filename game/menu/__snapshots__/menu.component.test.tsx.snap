// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game / Menu / Component > should match snapshot 1`] = `
<div>
  <div
    style="display: block;"
  >
    <nav
      class="navigation_navigation__hx5tsf0"
    >
      <a
        href="/"
      >
        Game
      </a>
      <a
        href="/blog"
      >
        Blog
      </a>
      <a
        href="/wiki"
      >
        Wiki
      </a>
    </nav>
    <div
      class="menu__1ofgcy10"
    >
      <button
        class="menu__1ofgcy11"
        id="play-button"
      >
        <h3>
          Click to play
        </h3>
      </button>
    </div>
  </div>
</div>
`;
